From 62c0ef7f809b6007929f3ff16588295b0f9daa0a Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 06 May 2020 22:23:57 +0300
Subject: [PATCH] [SW-10690] cmd: pci: add init of flr

Since in Linux we start handling the FLR by moving to H/W FLR handling
and performing a hard reset, we need here to hand it back to to S/W.
Add module_pci_setup() function which performs this action by device
specific flows.

Change-Id: I5242400ec718a549669772ca3d412c6f4b1b6880
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/app/preboot.c b/app/preboot.c
index e2d935a..b2bfbf9 100644
--- a/app/preboot.c
+++ b/app/preboot.c
@@ -27,7 +27,7 @@
 
 static struct cmd_module_ops cmd_ops[] = {
 	{	/* PCI module */
-		.setup = noop,
+		.setup = module_pci_setup,
 		.check = module_pci_check,
 		.disable = noop,
 		.run = noop,
diff --git a/common/cmd_module.h b/common/cmd_module.h
index 1822777..be1dcd2 100644
--- a/common/cmd_module.h
+++ b/common/cmd_module.h
@@ -63,8 +63,10 @@
 static inline void noop(void) {}
 
 /* PCI */
+void module_pci_setup(void);
 enum cmd_check_res module_pci_check(void);
 void module_pci_run(void);
+void pci_dev_specific_setup(void);
 
 /* FWUPD */
 void module_fwupd_setup(void);
diff --git a/common/module_pci.c b/common/module_pci.c
index 1443d33..0060664 100644
--- a/common/module_pci.c
+++ b/common/module_pci.c
@@ -22,6 +22,13 @@
 	.hot_rst_reg_val = ~1,
 };
 
+__attribute__((weak)) void pci_dev_specific_setup(void) {}
+
+void module_pci_setup(void)
+{
+	pci_dev_specific_setup();
+}
+
 enum cmd_check_res module_pci_check(void)
 {
 	pci_sts.flr_reg_val = regmap_readl(mmPCIE_AUX_FLR_INT);
diff --git a/gaudi/common.c b/gaudi/common.c
index fc08060..a5b44af 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -241,3 +241,9 @@
 
 	spi_copy((void *)loadaddr, offset, size, name);
 }
+
+void pci_dev_specific_setup(void)
+{
+	/* to be handled by SW - enable */
+	regmap_writel(0, mmPCIE_AUX_FLR_CTRL);
+}
diff --git a/goya2/common.c b/goya2/common.c
index 2895c0d..050cd08 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -255,3 +255,9 @@
 
 	spi_copy((void *)loadaddr, offset, size, name);
 }
+
+void pci_dev_specific_setup(void)
+{
+	/* to be handled by SW - enable */
+	regmap_writel(0, mmPCIE_AUX_FLR_CTRL);
+}
